Lib/urllib/request.py

```diff
@@ -950,27 +957,51 @@ def __init__(self, password_mgr=None):
         self.passwd = password_mgr
         self.add_password = self.passwd.add_password
 
+    def _parse_realm(self, header):
+        # parse WWW-Authenticate header: accept multiple challenges per header
+        found_challenge = False
+        for mo in AbstractBasicAuthHandler.rx.finditer(header):
+            scheme, quote, realm = mo.groups()
+            if quote not in ['"',"'"]:
+                warnings.warn("Basic Auth Realm was unquoted",
+                              UserWarning, 2)
+
+            yield (scheme, realm)
+
+            found_challenge = True
+
+        if not found_challenge:
+            if header:
+                scheme = header.split()[0]
+            else:
+                scheme = ''
+            yield (scheme, None)
+
     def http_error_auth_reqed(self, authreq, host, req, headers):
         # host may be an authority (without userinfo) or a URL with an
         # authority
-        # XXX could be multiple headers
-        authreq = headers.get(authreq, None)
+        headers = headers.get_all(authreq)
+        if not headers:
```

Comment:
get_all() returns None if no header is found. We cannot iterate on None.

I didn't want to try to pass a second argument to get_all(), since I didn't want to check if it's accepted in all Python versions (since this security fix should be applied to many Python versions).

I prefer an explicit test on the result which should be safe.